src/Database/Esqueleto/Experimental.hs

```diff
@@ -0,0 +1,845 @@
+-- type is defined as a binary tree, with @SelectQuery@ on the leaves.
+--
+-- Each constructor corresponding to the aforementioned set operations
+-- can be used as an infix function in a @from@ to help with readability
+-- and lead to code that closely resembles the underlying SQL. For example,
+--
+-- @
+-- select $ from $
+--   (SelectQuery ...)
+--   \`Union\`
+--   (SelectQuery ...)
+-- @
+--
+-- is translated into
+--
+-- @
+-- SELECT * FROM (
+--   (SELECT * FROM ...)
+--   UNION
+--   (SELECT * FROM ...)
+-- )
+-- @
+--
+-- @SelectQuery@ can be used without any of the set operations to construct
+-- a subquery. This can be used in 'JOIN' trees. For example,
+--
+-- @
+-- select $ from $
+--   Table @SomeTable
+--   \`InnerJoin\` (SelectQuery ...)
+--   \`on\` ...
+-- @
+--
+-- is translated into
+--
+-- @
+-- SELECT *
+-- FROM SomeTable
+-- INNER JOIN (SELECT * FROM ...)
+-- ON ...
+-- @
+data SqlSetOperation a =
+    Union (SqlSetOperation a) (SqlSetOperation a)
+  | UnionAll (SqlSetOperation a) (SqlSetOperation a)
+  | Except (SqlSetOperation a) (SqlSetOperation a)
+  | Intersect (SqlSetOperation a) (SqlSetOperation a)
+  | SelectQuery (SqlQuery a)
+
+-- | Data type that represents the syntax of a 'JOIN' tree. In practice,
+-- only the @Table@ constructor is used directly when writing queries. For example,
+--
+-- @
+-- select $ from $ Table @People
+-- @
+data From a where
+  Table         :: PersistEntity ent => From (SqlExpr (Entity ent))
+  SubQuery      :: (SqlSelect a' r, SqlSelect a'' r', ToAlias a, a' ~ ToAliasT a, ToAliasReference a', ToAliasReferenceT a' ~ a'')
+                => SqlQuery a
+                -> From a''
+  SqlSetOperation :: (SqlSelect a' r, ToAlias a, a' ~ ToAliasT a, ToAliasReference a', ToAliasReferenceT a' ~ a'')
+                  => SqlSetOperation a
+                  -> From a''
+  InnerJoinFrom :: From a
+                -> (From b, (a :& b) -> SqlExpr (Value Bool))
+                -> From (a :& b)
+  CrossJoinFrom :: From a
+                -> From b
+                -> From (a :& b)
+  LeftJoinFrom  :: ToMaybe b
+                => From a
+                -> (From b, (a :& ToMaybeT b) -> SqlExpr (Value Bool))
+                -> From (a :& ToMaybeT b)
+  RightJoinFrom :: ToMaybe a
+                => From a
+                -> (From b, (ToMaybeT a :& b) -> SqlExpr (Value Bool))
+                -> From (ToMaybeT a :& b)
+  FullJoinFrom  :: (ToMaybe a, ToMaybe b )
+                => From a
+                -> (From b, (ToMaybeT a :& ToMaybeT b) -> SqlExpr (Value Bool))
+                -> From (ToMaybeT a :& ToMaybeT b)
+
+-- | An @ON@ clause that describes how two tables are related. This should be
+-- used as an infix operator after a 'JOIN'. For example,
+--
+-- @
+-- select $
+-- from $ Table @Person
+-- \`InnerJoin\` Table @BlogPost
+-- \`on\` (\\(p :& bP) ->
+--         p ^. PersonId ==. bP ^. BlogPostAuthorId)
+-- @
+--
+on :: ToFrom a => a -> b -> (a, b)
```

Comment:
```suggestion
on :: ToFrom a => a -> (b -> SqlExpr (Value Bool)) -> (a, b -> SqlExpr (Value Bool))
```

I've found that locally this improves the type inference massively.

![image](https://user-images.githubusercontent.com/7310112/77808907-f64b9880-7052-11ea-912f-cb36e25a95ab.png)


![image](https://user-images.githubusercontent.com/7310112/77808894-e5028c00-7052-11ea-8867-6f2ddd52fb0d.png)